src/main/java/com/segment/analytics/android/integrations/firebase/FirebaseIntegration.java

```diff
@@ -3,88 +3,242 @@
+    eventMapper.put("Product Shared", Event.SHARE);
+    eventMapper.put("Product Clicked", Event.SELECT_CONTENT);
+    eventMapper.put("Product Searched", Event.SEARCH);
+    eventMapper.put("Promotion Viewed", Event.PRESENT_OFFER);
+    return eventMapper;
+  }
 
-      Context context = analytics.getApplication();
+  private static final Map<String, String> propertyMapper = createPropertyMap();
+
+  private static Map<String, String> createPropertyMap() {
+    Map<String, String> propertyMapper = new HashMap<>();
+    propertyMapper.put("category", Param.ITEM_CATEGORY);
+    propertyMapper.put("product_id", Param.ITEM_ID);
+    propertyMapper.put("name", Param.ITEM_NAME);
+    propertyMapper.put("price", Param.PRICE);
+    propertyMapper.put("quantity", Param.QUANTITY);
+    propertyMapper.put("query", Param.SEARCH_TERM);
+    propertyMapper.put("shipping", Param.SHIPPING);
+    propertyMapper.put("tax", Param.TAX);
+    propertyMapper.put("total", Param.VALUE);
+    propertyMapper.put("revenue", Param.VALUE);
+    propertyMapper.put("order_id", Param.TRANSACTION_ID);
+    propertyMapper.put("currency", Param.CURRENCY);
+    return propertyMapper;
+  }
 
-      return new FirebaseIntegration(context);
-    }
+  public FirebaseIntegration(Context context, Logger logger) {
+    this.logger = logger;
+    firebaseAnalytics = FirebaseAnalytics.getInstance(context);
+  }
 
-    @Override public String key() {
-      return FIREBASE_ANALYTICS_KEY;
+  @Override
+  public void onActivityResumed(Activity activity) {
+    super.onActivityResumed(activity);
+
+    PackageManager packageManager = activity.getPackageManager();
+    try {
+      ActivityInfo info =
+          packageManager.getActivityInfo(activity.getComponentName(), PackageManager.GET_META_DATA);
+      String activityLabel = info.loadLabel(packageManager).toString();
+      firebaseAnalytics.setCurrentScreen(activity, activityLabel, null);
+      logger.verbose(
+          "firebaseAnalytics.setCurrentScreen(activity, %s, null);",
+          activityLabel);
+    } catch (PackageManager.NameNotFoundException e) {
+      throw new AssertionError("Activity Not Found: " + e.toString());
     }
-  };
-
-  private static final String FIREBASE_ANALYTICS_KEY = "Firebase Analytics";
-
-  public FirebaseIntegration(Context context) {
-    mFirebaseAnalytics = FirebaseAnalytics.getInstance(context);
   }
 
-  @Override public void onActivityStarted(Activity activity) {
-    super.onActivityStarted(activity);
+  @Override
+  public void identify(IdentifyPayload identify) {
+    super.identify(identify);
 
+    if (!isNullOrEmpty(identify.userId())) {
+      firebaseAnalytics.setUserId(identify.userId());
+    }
+    Map<String, Object> traits = identify.traits();
+    for (Map.Entry<String, Object> entry : traits.entrySet()) {
+      String trait = entry.getKey();
+      trait = trait.trim().replaceAll(" ", "_");
+      if (trait.length() > 40) {
```

Comment:
This check is unnecessary like the earlier `trait.contains(" ")` check.

Thought this will be also addressed if you move `makeKey` into a utility function with the implementation I suggested.